messaging/tck/src/main/java/org/eclipse/microprofile/reactive/messaging/tck/arquillian/TckArquillianDeployListener.java

```diff
@@ -0,0 +1,141 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+package org.eclipse.microprofile.reactive.messaging.tck.arquillian;
+
+import org.eclipse.microprofile.reactive.messaging.tck.ContainerController;
+import org.eclipse.microprofile.reactive.messaging.tck.MockedReceiver;
+import org.eclipse.microprofile.reactive.messaging.tck.MockedSender;
+import org.eclipse.microprofile.reactive.messaging.tck.ReceiveTimeoutException;
+import org.eclipse.microprofile.reactive.messaging.tck.TckMessagingManager;
+import org.eclipse.microprofile.reactive.messaging.tck.spi.TestEnvironment;
+import org.eclipse.microprofile.reactive.messaging.tck.mocks.MockPayload;
+import org.eclipse.microprofile.reactive.messaging.tck.mocks.SimpleMessage;
+import org.eclipse.microprofile.reactive.messaging.tck.spi.TckContainer;
+import org.jboss.arquillian.container.spi.client.container.DeployableContainer;
+import org.jboss.arquillian.container.spi.client.container.DeploymentException;
+import org.jboss.arquillian.container.spi.client.deployment.DeploymentDescription;
+import org.jboss.arquillian.container.spi.event.container.AfterUnDeploy;
+import org.jboss.arquillian.container.spi.event.container.BeforeDeploy;
+import org.jboss.arquillian.core.api.annotation.Observes;
+import org.jboss.arquillian.test.spi.TestClass;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class TckArquillianDeployListener {
+
+  private final TckContainer tckContainer;
+
+  private final Map<Class<?>, List<DeploymentDescription>> activeDeployments = new ConcurrentHashMap<>();
+
+  /**
+   * I don't know how deployments work in arquillian, I assume I need to explicitly deploy everything?
```

Comment:
you need to create the JAR/WAR file you want to deploy into the container, so it is not different. The only thing is that in production code all package is done by build tool and here you need to do it programmatically, but it is exactly the same.

Reply:
Yes and no. As I understand it, you don't need to explicitly package the test class, that's done automatically, at least as I understand from the docs. But I don't know how it's done, and what about the classes that the test class depends on? Is there a remote classloader that points back to this classloader? So I'm not sure if I need everything here because if there is a remote classloader, won't it automatically load some of these things?